collector-sqs/src/main/java/zipkin/collector/awssqs/AwsSqsSpanProcessor.java

```diff
@@ -0,0 +1,155 @@
+  private final String queueUrl;
+  private final int waitTimeSeconds;
+  private final AtomicBoolean closed = new AtomicBoolean(false);
+  private final AtomicReference<CheckResult> status = new AtomicReference<>(CheckResult.OK);
+
+  AwsSqsSpanProcessor(AmazonSQSAsync client, Collector collector, String queueUrl,
+      int waitTimeSeconds) {
+    this.client = client;
+    this.collector = collector;
+    this.queueUrl = queueUrl;
+    this.waitTimeSeconds = waitTimeSeconds;
+  }
+
+  @Override public CheckResult check() {
+    return status.get();
+  }
+
+  @Override public void close() throws IOException {
+    if (!closed.getAndSet(true)) {
+      client.shutdown();
+    }
+  }
+
+  AwsSqsSpanProcessor run() {
+    final ReceiveMessageRequest request = new ReceiveMessageRequest(queueUrl)
+        .withWaitTimeSeconds(waitTimeSeconds)
+        .withMessageAttributeNames("spans");
+
+    if (!closed.get()) {
+      client.receiveMessageAsync(request, new Handler(this));
+    }
+
+    return this;
+  }
+
+  private void process(final List<Message> messages) {
+    final List<DeleteMessageBatchRequestEntry> deletes = new LinkedList<>();
+    for (Message message : messages) {
+      MessageAttributeValue spanAttributes = message.getMessageAttributes().get("spans");
+      if (spanAttributes != null) {
+        final List<Span> spans = new LinkedList<>();
+        byte[] bytes;
+        switch (spanAttributes.getDataType()) {
+          case "Binary.JSON":
+            bytes = spanAttributes.getBinaryValue().array();
+            spans.addAll(Codec.JSON.readSpans(bytes));
+            break;
+          case "Binary.THRIFT":
+            bytes = spanAttributes.getBinaryValue().array();
+            spans.addAll(Codec.THRIFT.readSpans(bytes));
+            break;
+          default:
+            log.error(String.format("Unsupported codec %s", spanAttributes.getDataType()));
+            continue;
+        }
+
+        collector.accept(spans, new CollectorCallback(spans));
+
+        deletes.add(
+            new DeleteMessageBatchRequestEntry(message.getMessageId(), message.getReceiptHandle()));
+      }
+    }
+
+    client.deleteMessageBatch(queueUrl, deletes);
+
+  }
+
+  private static final class Handler implements
+      AsyncHandler<ReceiveMessageRequest, ReceiveMessageResult> {
+
+    private final AwsSqsSpanProcessor processor;
+
+    Handler(AwsSqsSpanProcessor processor) {
+      this.processor = processor;
+    }
+
+    @Override public void onError(Exception exception) {
+      // TODO when should this just give up?
+      processor.status.set(CheckResult.failed(exception));
+      processor.run();
+    }
+
+    @Override
+    public void onSuccess(ReceiveMessageRequest request, ReceiveMessageResult result) {
+      processor.process(result.getMessages());
+      processor.status.lazySet(CheckResult.OK);
+      processor.run();
+    }
+  }
+
+  private static final class CollectorCallback implements Callback<Void> {
```

Comment:
if you notice zipkin.collector.Collector has utilities to keep track of things like this